### ISSUE TITLE
RCA for array inside a tuple across callable boundary always treated as dynamically sized

### ISSUE THREAD

**Issue by swernli** [microsoft/qsharp#1391]
With current RCA behaviors, all items inside a tuple are assumed to be dynamic, which for arrays means they are treated as totally dynamic and do not recognize when a static length is used.  This can make certain code patterns be treated as if they are using dynamically sized arrays when technically they are not. For example:
```qsharp
@EntryPoint()
operation Main() : (Result, Bool[]) {
    let q = Qubit();
    let r = MResetZ(q);
    (r, [r == Zero, r == One])
}
```
This is an expected limitation of how tuples treat their contents in RCA, and may be difficult to account for. We should investigate whether some technique could allow for statically sized arrays to be used in a dynamic tuple context.

**Comment by cesarzc** [comment]
This is addressed by #1454.

However, I think we should keep it open until we make the change to remove `ValueKind` and just have `RuntimeKind` (to simplify the RCA implementation). What do you think @swernli?

**Comment by swernli** [comment]
> This is addressed by #1454.
> 
> However, I think we should keep it open until we make the change to remove `ValueKind` and just have `RuntimeKind` (to simplify the RCA implementation). What do you think @swernli?

I'm on the fence... I could see closing this and opening a more scoped one for the possible internal refactor. The immediate issue from the title has been resolved, so the future update would have no customer visible change.
